cropper/src/main/java/com/canhub/cropper/CropImageIntentChooser.kt

```diff
@@ -0,0 +1,223 @@
+        val intents: MutableList<Intent> = ArrayList()
+        val galleryIntent = if (action == Intent.ACTION_GET_CONTENT) Intent(action)
+        else Intent(action, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
+        galleryIntent.type = "image/*"
+        val listGallery = packageManager.queryIntentActivities(galleryIntent, 0)
+        for (res in listGallery) {
+            val intent = Intent(galleryIntent)
+            intent.component = ComponentName(res.activityInfo.packageName, res.activityInfo.name)
+            intent.setPackage(res.activityInfo.packageName)
+            intents.add(intent)
+        }
+        // sort intents
+        val priorityIntents = mutableListOf<Intent>()
+        for (pkgName in priorityIntentList) {
+            intents.firstOrNull { it.`package` == pkgName }?.let {
+                intents.remove(it)
+                priorityIntents.add(it)
+            }
+        }
+        intents.addAll(0, priorityIntents)
+        return intents
+    }
+
+    /**
+     * Check if explicetly requesting camera permission is required.<br></br>
+     * It is required in Android Marshmellow and above if "CAMERA" permission is requested in the
+     * manifest.<br></br>
+     * See [StackOverflow
+     * question](http://stackoverflow.com/questions/32789027/android-m-camera-intent-permission-bug).
+     */
+    private fun isExplicitCameraPermissionRequired(context: Context): Boolean {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
+            hasCameraPermissionInManifest(context) &&
+            context.checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED
+    }
+
+    /**
+     * Check if the app requests a specific permission in the manifest.
+     *
+     * @param context the context of your activity to check for permissions
+     * @return true - the permission in requested in manifest, false - not.
+     */
+    private fun hasCameraPermissionInManifest(context: Context): Boolean {
+        val packageName = context.packageName
+        try {
+            val packageInfo =
+                context.packageManager.getPackageInfo(packageName, PackageManager.GET_PERMISSIONS)
+            val declaredPermissions = packageInfo.requestedPermissions
+            if (
+                declaredPermissions
+                    ?.any {
+                        it?.equals("android.permission.CAMERA", true) == true
+                    } == true
+            ) {
+                return true
+            }
+        } catch (e: PackageManager.NameNotFoundException) {
+            // Since the package name cannot be found we return false below
+            // because this means that the camera permission hasn't been declared
+            // by the user for this package so we can't show the camera app among
+            // among the list of apps
+            e.printStackTrace()
+        }
+        return false
+    }
```

Comment:
What id the developer, is using the camera activity contract that doesn't need permission anymore? 🤔

Reply:
When we try to display the Camera app as a list in the intent chooser it seems to require that the developers have the camera permission declared in their Manifest file. I've tried testing it without but can't get it working without that.

I'm guessing this is because it's called by the intent choose and not directly by the user.

I'm not sure if I'm testing it wrong.

Reply:
But if they don't declare it, it will still show the intent chooser just without the camera app

Reply:
In this case we should say this on the documentation for turning this feature on

Reply:
Alright I'm adding it